HULKs/hulk
# Aliveness service and pepsi subcommand

## Introduced Changes

This PR adds an aliveness service for querying multiple status information from NAOs in the network. The aliveness servive is built with yocto and will be integrated in the upcoming HULKs-OS version (see HULKs/meta-hulks#10).

Additionally, an `aliveness` subcommand is added to pepsi.
By default, only irregular information like non-active services, outdated HULKs-OS versions and battery charge levels below 95% are displayed. Using `-v`/`--verbose` or `-j`/`--json`, the user can retrieve all information available via aliveness in either a human- or machine-readable format.

The user can furthermore set a timeout via `-t`/`--timeout` (defaulting to 200ms) and specify NAO addresses (e.g. `22` or `10.1.24.22`) for querying the aliveness information only from specific NAOs.

## ToDo / Known Issues

The beacon for querying the aliveness information is send via multicast and the answers are received via unicast.
Since the answers are from a different IP address, most firewalls will block them.

Hence, the user has to allow the incoming messages in their firewall settings, e.g. for ufw by adding the following rule:

```
ufw allow proto udp from 10.1.24.0/24
```

## Ideas for Next Iterations (Not This PR)

Dynamic tab completions for pepsi subcommands such as `pepsi upload` and `pepsi poweroff` using the aliveness information are already WIP, but have been factored out to a different PR.

## How to Test

First, execute `cargo build` in `tools/aliveness` after sourcing the SDK.
Copy the target to a NAO, ssh into it and execute the target.
Now you can try the pepsi aliveness subcommand with any of the options, e.g.

```
./pepsi aliveness
./pepsi aliveness 27 32
./pepsi aliveness --json
./pepsi aliveness --timeout 500 -v
```

Make sure your firewall allows the incoming responses as described above if you don't see any responses.
